File: python/creational/factory_method.py
class Car:
    @staticmethod
    def my_vehicle():
        print('your vehicle is car')


class Motorbike:
    @staticmethod
    def my_vehicle():
        print('your vehicle is motorbike')


def use_vehicle(vehicle):
    match vehicle:
        case 'motorbike':
            return Motorbike()
        case 'car':
            return Car()
        case _:
            raise ValueError('Only motorbike or car are allowed')

motorbike = use_vehicle('motorbike')
car = use_vehicle('car')

motorbike.my_vehicle()
car.my_vehicle()

File: js/creational/factory_method.js
class Car {
    my_vehicle() {
        console.log('your vehicle is car')
    }
}

class Motorbike {
    my_vehicle() {
        console.log('your vehicle is motorbike')
    }
}

const useVehicle = (vehicle) => {
    switch (vehicle) {
        case 'motorbike':
            return new Car()
        case 'car':
            return new Motorbike()
        default:
            throw new Error('It allows to use motorbike or car')
    }
}

const motorbike = useVehicle('motorbike');
const car = useVehicle('car')

motorbike.my_vehicle()
car.my_vehicle()
